buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -222,256 +289,278 @@ private int usage(int freeBytes) {
             // This is need as we may add it back and so alter the linked-list structure.
-            PoolSubpage<T> head = arena.findSubpagePoolHead(subpage.elemSize);
             synchronized (head) {
-                if (subpage.free(head, bitmapIdx & 0x3FFFFFFF)) {
+                if (subpage.free(head, bitmapIdx(handle))) {
+                    //the subpage is still used, do not free it
                     return;
                 }
             }
         }
-        freeBytes += runLength(memoryMapIdx);
-        setValue(memoryMapIdx, depth(memoryMapIdx));
-        updateParentsFree(memoryMapIdx);
 
-        if (nioBuffer != null && cachedNioBuffers != null &&
-                cachedNioBuffers.size() < PooledByteBufAllocator.DEFAULT_MAX_CACHED_BYTEBUFFERS_PER_CHUNK) {
-            cachedNioBuffers.offer(nioBuffer);
+        //start free run
+        int pages = runPages(handle);
+
+        synchronized (runsAvail) {
+            // collapse continuous runs, successfully collapsed runs
+            // will be removed from runsAvail and runsAvailMap
+            long finalRun = collapseRuns(handle);
+
+            //set run as not used
+            finalRun &= ~(1L << ISUSED_SHIFT);
+            //if it is a subpage, set it to run
+            finalRun &= ~(1L << ISSUBPAGE_SHIFT);
+
+            insertAvailRun(runOffset(finalRun), runPages(finalRun), finalRun);
+            freeBytes += pages << pageShifts;
         }
-    }
 
-    void initBuf(PooledByteBuf<T> buf, ByteBuffer nioBuffer, long handle, int reqCapacity,
-                 PoolThreadCache threadCache) {
-        int memoryMapIdx = memoryMapIdx(handle);
-        int bitmapIdx = bitmapIdx(handle);
-        if (bitmapIdx == 0) {
-            byte val = value(memoryMapIdx);
-            assert val == unusable : String.valueOf(val);
-            buf.init(this, nioBuffer, handle, runOffset(memoryMapIdx) + offset,
-                    reqCapacity, runLength(memoryMapIdx), threadCache);
-        } else {
-            initBufWithSubpage(buf, nioBuffer, handle, bitmapIdx, reqCapacity, threadCache);
+        if (nioBuffer != null && cachedNioBuffers != null &&
+            cachedNioBuffers.size() < PooledByteBufAllocator.DEFAULT_MAX_CACHED_BYTEBUFFERS_PER_CHUNK) {
+            cachedNioBuffers.offer(nioBuffer);
         }
     }
 
-    void initBufWithSubpage(PooledByteBuf<T> buf, ByteBuffer nioBuffer, long handle, int reqCapacity,
-                            PoolThreadCache threadCache) {
-        initBufWithSubpage(buf, nioBuffer, handle, bitmapIdx(handle), reqCapacity, threadCache);
+    private long collapseRuns(long handle) {
+        return collapseNext(collapsePast(handle));
     }
 
-    private void initBufWithSubpage(PooledByteBuf<T> buf, ByteBuffer nioBuffer,
-                                    long handle, int bitmapIdx, int reqCapacity, PoolThreadCache threadCache) {
-        assert bitmapIdx != 0;
+    private long collapsePast(long handle) {
+        while (true) {
```

Comment:
in Netty we usually use a "crying for loop" :) `for (;;) {....}`. Consider changing this one here to make it consistent